platform/mbed_enum_flags.h

```diff
@@ -0,0 +1,158 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_SCOPED_ENUM_FLAGS_H
+#define MBED_SCOPED_ENUM_FLAGS_H
+
+#ifdef __cplusplus
+#include <type_traits>
+
+#define ENUM_FLAG_OPERATOR(T,OP) \
+inline constexpr T operator OP( T lhs, T rhs) {\
+    return ( T ) ( \
+        static_cast<std::underlying_type< T >::type>(lhs) OP \
+        static_cast<std::underlying_type< T >::type>(rhs)); \
+};
+
+
+#define ENUM_FLAG_OPERATOR_LHS_REFERNCE(T, X) \
+inline T & operator X (T & a, T b) { \
+    a = a X b; \
+    return a; \
+}
+
+
+/**
+ * @brief Applies bitwise operators to a enum class defined elsewhere.
+ *
+ * @param T             The enum class typename
+ *
+ * This macro applies the bitwise negate, AND, OR, XOR operators and the
+ * corresponding assignment operators to an existing enum class. The macro
+ * uses underlying type traits to convert back and forth.
+ *
+ * Due to the way the macro works, there is no semicolon required at the end.
+ *
+ * Usage:
+ * @code
+ * external_enum.h:
+ *
+ * enum class SpokenLanguages : uint8_t {
+ *     Sindarin    = (1 << 0),
+ *     Westron     = (1 << 1),
+ *     Rohirric    = (1 << 2),
+ *     BlackSpeech = (1 << 3)
+ * };
+ *
+ * my_code.cpp:
+ *
+ * #include "mbed_enum_flags.h"
+ * #include "external_enum.h"
+ *
+ * MBED_ENUM_FLAG_APPLY_OPERATORS(SpokenLanguages)
+ *
+ * SpokenLanguages gandalf = SpokenLanguages::Sindarin | SpokenLanguages::Westron |
+ *     SpokenLanguages::Rohirric | SpokenLanguages::BlackSpeech;
+ * @endcode
+ *
+ */
+#define MBED_ENUM_FLAG_APPLY_OPERATORS(T) \
+\
+inline constexpr T operator ~( T lhs) { \
+    return ( T ) (~static_cast<std::underlying_type< T >::type>(lhs)) ;\
+}\
+ENUM_FLAG_OPERATOR(T, |)\
+ENUM_FLAG_OPERATOR(T, ^)\
+ENUM_FLAG_OPERATOR(T, &)\
+ENUM_FLAG_OPERATOR_LHS_REFERNCE(T, |=) \
+ENUM_FLAG_OPERATOR_LHS_REFERNCE(T, ^=) \
+ENUM_FLAG_OPERATOR_LHS_REFERNCE(T, &=) \
+
+
+#define MBED_SCOPED_ENUM_FLAGS_TYPED(T, UNDER_TYPE) \
+enum class T : UNDER_TYPE; \
+MBED_ENUM_FLAG_APPLY_OPERATORS(T) \
+enum class T : UNDER_TYPE
+
+#define MBED_SCOPED_ENUM_FLAGS_UNTYPED(T) \
+enum class T; \
+MBED_ENUM_FLAG_APPLY_OPERATORS(T) \
+enum class T
+
+#define MBED_SCOPED_ENUM_FLAGS_CHOOSER(_1, _2, NAME, ...) NAME
+
+/**
+ * @brief Creates an enum class with bitwise operator overloads.
+ *
+ * @param T             The enum class typename
+ * @param UNDER_TYPE    Optional: specify the underlying integral type. If left blank will
+ *                      default to the compiler default (by standard, an int)
```

Comment:
I attempted to base this on the definition in the documents of what the `enum class` underlying type default was because I believed this was in a reference somewhere. That being said, I think I overstated it for two reasons:
1.  The standard says it is an `int` (however many bits that actually is) unless the compiler notices it needs to be bigger. At least I thought it was so.
1. It is possible to turn on compiler flags to make this smaller. This doesn't actually force an `int`, it simply leaves it to the compiler default. Not that this isn't potentially fraught will peril, but _caveat programmator_.

So, I will try not to state what I do not actually know to be true.

Reply:
> unless the compiler notices it needs to be bigger

That's not so, unless I'm missing something. Non-scoped enum is an implementation-defined type by default, possibly smaller than int, and whether it can be bigger than int is implementation-defined.  (IIRC). Scoped enum is just `int` unless explicitly typed.

> It is possible to turn on compiler flags to make this smaller. 

Are you sure? There are flags which control the implementation-defined behaviour for plain `enum`, but I would doubt they change standard-conformant behaviour for scoped enum.

But there is an important point that you shouldn't "restate" defaults in code. If you think you know what the lower layer's default is, but you don't care, you should let it do its default, rather than specifying it explicitly. That leaves the door open to some sort of environment config change.